universum/modules/vcs/vcs.py

```diff
@@ -172,12 +172,7 @@ def clean_sources_silently(self):
         except OSError:
             pass
 
-    def copy_cl_files_and_revert(self) -> None:
-        raise NotImplementedError("Not applicable")
-
     @make_block("Revert repository")
     def revert_repository(self) -> Optional[List[Tuple[Optional[str], Optional[str], Optional[str]]]]:
-        if isinstance(self.driver, base_vcs.BaseDownloadVcs):
-            diff = self.driver.copy_cl_files_and_revert()
-            return diff
-        raise NotImplementedError("The driver is unable to perform the requested action")
+        diff = self.driver.copy_cl_files_and_revert()
```

Comment:
Do I understand it correctly: mypy doesn't know that driver has this function, so we had to move it to BaseVcs?

Reply:
driver can be either DownloadVcs, SubmitVcs, or PollVcs - but only DownloadVcs has 'copy_cl_files_and_revert'.
This causes an issue in case this method is called for non-applicable driver.
By moving 'copy_cl_files_and_revert' to BaseVcs (common parent), we avoid the issue.

Reply:
Of course - this is not the only way to resolve this. Checking whether the driver is fitting class also works - but that is less pythonic

Reply:
or we can always silence the mypy warning

Reply:
I think the real issue is incorrect annotation of `driver` member if `MainVcs`.
`driver` cannot be SybmitVcs or PollVcs.

Reply:
Got it. Fixed